In [ ]:
import pandas as pd
import numpy as np 

import matplotlib.pyplot as plt
import seaborn as sns 

# Settings (very important for clean output)
sns.set(style="whitegrid")  #Sets default grid for all my plots
plt.rcParams['figure.figsize'] = (10, 6) #Sets default size of all plots (W by H)

import warnings
warnings.filterwarnings("ignore")

In [12]:
import os 
os.getcwd()


'd:\\credit-risk-scoring\\notebooks'

In [13]:
os.listdir()


['EDA.ipynb']

In [14]:
os.listdir('../data/raw')

['accepted_2007_to_2018Q4.csv', 'rejected_2007_to_2018Q4.csv']

In [3]:
pd.set_option('display.max_columns', None)

In [15]:
df = pd.read_csv('../data/raw/accepted_2007_to_2018Q4.csv')

In [ ]:

# # Define paths correctly relative to the notebooks directory
# csv_path = '../data/raw/accepted_2007_to_2018Q4.csv'
# parquet_path = '../data/raw/accepted_2007_to_2018Q4.parquet'

# # Read the CSV (this will be slow the final time)
# print(f"Reading CSV from {csv_path}...")
# df_csv = pd.read_csv(csv_path, low_memory=False)

# # Write to Parquet using the pyarrow engine
# print(f"Saving to Parquet at {parquet_path}...")
# df_csv.to_parquet(parquet_path, engine='pyarrow', index=False)

# print("Conversion complete.")


Reading CSV from ../data/raw/accepted_2007_to_2018Q4.csv...
Saving to Parquet at ../data/raw/accepted_2007_to_2018Q4.parquet...
Conversion complete.


In [ ]:
# Load the Parquet file (much faster and memory efficient)
parquet_path = '../data/raw/accepted_2007_to_2018Q4.parquet'

df = pd.read_parquet(parquet_path, engine='pyarrow')

# Verify the load
df.head()


,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,home_ownership,annual_inc,verification_status,issue_d,loan_status,pymnt_plan,url,desc,purpose,title,zip_code,addr_state,dti,delinq_2yrs,earliest_cr_line,fico_range_low,fico_range_high,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,initial_list_status,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_d,last_pymnt_amnt,next_pymnt_d,last_credit_pull_d,last_fico_range_high,last_fico_range_low,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,application_type,annual_inc_joint,dti_joint,verification_status_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_fico_range_low,sec_app_fico_range_high,sec_app_earliest_cr_line,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,hardship_flag,hardship_type,hardship_reason,hardship_status,deferral_term,hardship_amount,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_length,hardship_dpd,hardship_loan_status,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,68407277,NaN,3600.0,3600.0,3600.0,36 months,13.99,123.03,C,C4,leadman,10+ years,MORTGAGE,55000.0,Not Verified,Dec-2015,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,None,debt_consolidation,Debt consolidation,190xx,PA,5.91,0.0,Aug-2003,675.0,679.0,1.0,30.0,NaN,7.0,0.0,2765.0,29.7,13.0,w,0.00,0.00,4421.723917,4421.72,3600.00,821.72,0.0,0.0,0.0,Jan-2019,122.67,None,Mar-2019,564.0,560.0,0.0,30.0,1.0,Individual,NaN,NaN,None,0.0,722.0,144904.0,2.0,2.0,0.0,1.0,21.0,4981.0,36.0,3.0,3.0,722.0,34.0,9300.0,3.0,1.0,4.0,4.0,20701.0,1506.0,37.2,0.0,0.0,148.0,128.0,3.0,3.0,1.0,4.0,69.0,4.0,69.0,2.0,2.0,4.0,2.0,5.0,3.0,4.0,9.0,4.0,7.0,0.0,0.0,0.0,3.0,76.9,0.0,0.0,0.0,178050.0,7746.0,2400.0,13734.0,NaN,NaN,NaN,None,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,None,None,None,NaN,NaN,None,None,None,NaN,NaN,None,NaN,NaN,NaN,Cash,N,None,None,None,NaN,NaN,NaN
1,68355089,NaN,24700.0,24700.0,24700.0,36 months,11.99,820.28,C,C1,Engineer,10+ years,MORTGAGE,65000.0,Not Verified,Dec-2015,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,None,small_business,Business,577xx,SD,16.06,1.0,Dec-1999,715.0,719.0,4.0,6.0,NaN,22.0,0.0,21470.0,19.2,38.0,w,0.00,0.00,25679.660000,25679.66,24700.00,979.66,0.0,0.0,0.0,Jun-2016,926.35,None,Mar-2019,699.0,695.0,0.0,NaN,1.0,Individual,NaN,NaN,None,0.0,0.0,204396.0,1.0,1.0,0.0,1.0,19.0,18005.0,73.0,2.0,3.0,6472.0,29.0,111800.0,0.0,0.0,6.0,4.0,9733.0,57830.0,27.1,0.0,0.0,113.0,192.0,2.0,2.0,4.0,2.0,NaN,0.0,6.0,0.0,5.0,5.0,13.0,17.0,6.0,20.0,27.0,5.0,22.0,0.0,0.0,0.0,2.0,97.4,7.7,0.0,0.0,314017.0,39475.0,79300.0,24667.0,NaN,

In [ ]:
#Applying filter to 

df['issue_d'] = pd.to_datetime(df['issue_d'])
df = df[df['issue_d'].dt.year >= 2015]


#Wont have to work on massive data, spilt data as train on past data predict on future data, Train-2015-17, Test-2018

In [22]:
df.shape

(1794323, 151)

In [27]:
print(f"Dataset Shape:{df.shape[0]:,} rows and {df.shape[1]:} features ")

Dataset Shape:1,794,323 rows and 151 features 


In [35]:
# Inspecting data types
dtype_counts = df.dtypes.value_counts()
print(f"\nData Type Distribution:\n{dtype_counts}")


Data Type Distribution:
float64           113
object             37
datetime64[ns]      1
Name: count, dtype: int64


In [38]:
#Checking memory usages
memory_use = df.memory_usage(deep= True).sum() / 1024**2
print(f"\nMeomry usages:{memory_use:.2f}MB")


Meomry usages:4804.88MB


### The :.2f inside the print statement is a string formatting command in Python (used inside f-strings).

Here is exactly what it does:

 : tells Python, "I am about to specify how this variable should be formatted."
 .2 specifies the precision. It means, "Round this number to exactly 2 decimal places."
 f stands for float (a number with a decimal point). It tells Python to display the variable as a standard decimal number, rather than scientific notation.


### Why do we use it in this context?
The result of the memory calculation (df.memory_usage(deep=True).sum() / 1024**2) usually produces a very long floating-point number, something like 948.3847291048 MB.

Reporting a number like that is visually messy and distracting. By adding :.2f, we force Python to round and display it cleanly as 948.38 MB.

Examples to show the difference:

python
>value = 123.456789
>print(f"Normal: {value}")      # Outputs: Normal: 123.456789
>print(f"With .2f: {value:.2f}")  # Outputs: With .2f: 123.46
>print(f"With .0f: {value:.0f}")  # Outputs: With .0f: 123 (Rounds to whole number)
>In production and reporting (like Jupyter Notebooks meant for stakeholders), presenting clean, formatted numbers is always considered best practice!

In [39]:
# Random samples
df.sample(5, random_state=42)

,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,home_ownership,annual_inc,verification_status,issue_d,loan_status,pymnt_plan,url,desc,purpose,title,zip_code,addr_state,dti,delinq_2yrs,earliest_cr_line,fico_range_low,fico_range_high,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,initial_list_status,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_d,last_pymnt_amnt,next_pymnt_d,last_credit_pull_d,last_fico_range_high,last_fico_range_low,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,application_type,annual_inc_joint,dti_joint,verification_status_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_fico_range_low,sec_app_fico_range_high,sec_app_earliest_cr_line,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,hardship_flag,hardship_type,hardship_reason,hardship_status,deferral_term,hardship_amount,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_length,hardship_dpd,hardship_loan_status,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
1437501,142850764,NaN,3000.0,3000.0,3000.0,36 months,8.46,94.65,A,A5,Pilot,10+ years,MORTGAGE,135000.0,Not Verified,2018-10-01,Current,n,https://lendingclub.com/browse/loanDetail.acti...,None,major_purchase,Major purchase,363xx,AL,17.40,0.0,May-2006,705.0,709.0,2.0,39.0,NaN,12.0,0.0,5531.0,69.1,32.0,w,2702.88,2702.88,377.190000,377.19,297.12,80.07,0.0,0.0,0.0,Feb-2019,94.65,Apr-2019,Mar-2019,709.0,705.0,0.0,NaN,1.0,Individual,NaN,NaN,None,0.0,0.0,384588.0,3.0,9.0,5.0,8.0,3.0,237187.0,86.0,0.0,0.0,3079.0,86.0,8000.0,1.0,1.0,2.0,9.0,32049.0,421.0,88.0,0.0,0.0,149.0,132.0,45.0,3.0,5.0,119.0,NaN,3.0,NaN,0.0,1.0,2.0,1.0,1.0,24.0,2.0,3.0,2.0,12.0,0.0,0.0,0.0,5.0,93.8,100.0,0.0,0.0,418550.0,242718.0,3500.0,265550.0,NaN,NaN,NaN,None,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,None,None,None,NaN,NaN,None,None,None,NaN,NaN,None,NaN,NaN,NaN,Cash,N,None,None,None,NaN,NaN,NaN
910428,110665539,NaN,9600.0,9600.0,9600.0,36 months,9.93,309.45,B,B2,Public Relations Associate,5 years,RENT,31000.0,Not Verified,2017-06-01,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,None,debt_consolidation,Debt consolidation,440xx,OH,33.18,0.0,Sep-2006,770.0,774.0,1.0,NaN,NaN,16.0,0.0,4723.0,17.1,21.0,w,0.00,0.00,10272.926671,10272.93,9600.00,672.93,0.0,0.0,0.0,Mar-2018,7802.63,None,Dec-2018,769.0,765.0,0.0,NaN,1.0,Individual,NaN,NaN,None,0.0,0.0,36959.0,0.0,6.0,0.0,3.0,17.0,32236.0,60.0,0.0,1.0,1776.0,45.0,27700.0,2.0,0.0,1.0,4.0,2310.0,16077.0,22.7,0.0,0.0,129.0,58.0,16.0,16.0,0.0,16.0,NaN,5.0,NaN,0.0,4.0,4.0,7.0,7.0,12.0,9.0,9.0,4.0,16.0,0.0,0.0,0.0,0.0,100.0,0.0,0.0,0.0,81